core/kazoo_number_manager/src/carriers/knm_voxbone.erl

```diff
@@ -0,0 +1,332 @@
+                             ]),
+    {'ok', Response} = knm_voxbone_util:voxbone_request('put', <<"ordering/cart">>, [], Body),
+    Cart = kz_json:get_value(<<"cart">>, Response),
+    #{id => kz_json:get_value(<<"cartIdentifier">>, Cart), query_id => knm_search:query_id(Options)}.
+
+%%------------------------------------------------------------------------------
+%% @doc verify order can be fulfilled and initialize cart
+%% @end
+%%------------------------------------------------------------------------------
+-spec search(kz_term:api_pos_integer(), kz_term:proplist(), kz_term:proplist()) -> {'error', any()} | {'ok', knm_number:numbers()}.
+search(Quantity, Options, SearchOptions) ->
+    Balance = account_balance(),
+    #{total := Total, order :=  Order, cost := Cost} = do_search(Quantity, SearchOptions),
+    case {Total < Quantity, Balance}  of
+        %% We can't fulfill the quantity so return
+        {_, {'error', _}} -> knm_errors:unspecified("please contact the carrier administrator", -61);
+        {_, Balance} when Balance < Cost -> knm_errors:unspecified("please contact the carrier administrator", -61);
+        {'true', Balance} when Balance >= Cost -> knm_errors:unspecified("insufficient inventory to sastisfy request", 404);
+        {'false', Balance} when Balance >= Cost ->
+            Cart = create_cart(Options),
+            maybe_start_order(Cart, Order)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc get and process available DID quantities
+%% @end
+%%------------------------------------------------------------------------------
+-spec do_search(kz_term:api_pos_integer(), kz_term:proplist()) -> {'error', any()} | map().
+do_search(Quantity, SearchOptions) ->
+    QueryString = knm_voxbone_util:required_params() ++  SearchOptions,
+    case knm_voxbone_util:voxbone_request('get', <<"inventory/didgroup">>, QueryString) of
+        {'error', Reason} -> {'error',  Reason};
+        {'ok', DidGroups} -> process_search_results(Quantity, DidGroups)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc reserve quantities from DID groups and build order
+%% @end
+%%------------------------------------------------------------------------------
+-spec process_search_results(kz_term:api_pos_integer(), kz_json:object()) -> map().
+process_search_results(Quantity, Response) ->
+    DidGroups = kz_json:get_value(<<"didGroups">>, Response, kz_json:new()),
+    resolve_order_quantities(DidGroups, Quantity, #{total => 0, order => [], cost => 0}).
+
+%%------------------------------------------------------------------------------
+%% @doc reconcile a quantity of DIDs to purchase from a given DIDGroup
+%% @end
+%%------------------------------------------------------------------------------
+-spec resolve_order_quantities(kz_json:objects(), kz_term:api_pos_integer(), kz_term:proplist()) ->  map().
```

Comment:
Acc is not a proplist but a map right?

Reply:
Since 2nd clause can have Remaining = 0, spec should be `non_neg_integer() | 'undefined'` instead.